src/backend/distributed/executor/local_executor.c

```diff
@@ -585,6 +588,58 @@ ExecuteLocalTaskPlan(PlannedStmt *taskPlan, char *queryString,
 }
 
 
+/*
+ * RecordNonDistTableAccessesForTask records relation accesses for the non-distributed
+ * relations that given task will access.
+ */
+static void
+RecordNonDistTableAccessesForTask(Task *task)
+{
+	List *taskPlacementList = task->taskPlacementList;
+	int numberOfTaskPlacements = list_length(taskPlacementList);
+	if (numberOfTaskPlacements > 1)
+	{
+		/*
+		 * It can't be a non-distributed table, early exit. This is
+		 * because we are in the local executor and non-distributed
+		 * tables can't already have more than one placements in any node.
+		 */
+		return;
+	}
+
+	if (numberOfTaskPlacements == 0)
+	{
+		/*
+		 * FIXME: Unfortunately, it is possible due to
+		 * https://github.com/citusdata/citus/issues/4104.
+		 * We can safely remove this check when above bug is fixed.
+		 */
+		return;
+	}
+
+	ShardPlacement *taskPlacement = linitial(task->taskPlacementList);
+	uint64 shardId = taskPlacement->shardId;
+	if (shardId == INVALID_SHARD_ID)
+	{
+		/*
+		 * When a SELECT prunes down to 0 shard, we still may pass through
+		 * the local executor. In that case, we don't need to record any
+		 * relation access as we don't actually access any shard placement.
+		 */
+		return;
+	}
+
+	Oid relationId = RelationIdForShard(shardId);
```

Comment:
We  should use `task->relationShardList`  
 as  for  example `RecordRelationParallelSelectAccessForTask` does.

I'm actually confused by this change, aren't we already doing this change on #4139?

Reply:
yes, now we don't change local_executor in this pr, resolving